File: 프로그래머스/lv2/12909. 올바른 괄호/올바른 괄호.py
def solution(s):
    from collections import deque
    
    answer = False
    
    deq = deque()
    for i in range(len(s)):
        if len(s) % 2 == 1 :
            return False
        if s[i] == "(" :
            deq.append("(")
        elif "(" in deq and s[i] == ")":
            deq.remove("(")
        
    
    if not deq :
        return True
    if deq :
        return False
    

File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.java
import java.util.*;
class Solution {
    boolean solution(String s) {
        boolean answer = true;
        
        Deque<Character> deque = new ArrayDeque<>();
        
        if (s.length() % 2 == 1) {
            return false;
        }
        
        for(int i=0; i<s.length(); i++){
            char c = s.charAt(i);
            
            if(c == '('){
                deque.addFirst(c);
            }
            else{
                if (deque.isEmpty() || deque.peekFirst() != '(') {
                    return false;
                }
                deque.removeFirst();
            }
        }

        return deque.isEmpty();
    }
}